packages/walrus/src/client.ts

```diff
@@ -308,6 +306,91 @@ export class WalrusClient {
 		return currentEpoch;
 	}
 
+	async getBlobMetadata({ blobId, signal }: GetBlobMetadataOptions) {
+		const controller = new AbortController();
+		signal?.addEventListener('abort', () => {
+			controller.abort();
+		});
+
+		const committee = await this.#getReadCommittee({ blobId, signal });
+		const randomizedNodes = shuffle(committee.nodes);
+
+		let numNotFoundWeight = 0;
+		let numBlockedWeight = 0;
+
+		const executors = randomizedNodes.map((node) => {
+			return async () => {
+				try {
+					return await this.#storageNodeClient.getBlobMetadata(
+						{ blobId },
+						{ nodeUrl: node.networkUrl, signal: controller.signal },
+					);
+				} catch (error) {
+					if (error instanceof NotFoundError) {
+						numNotFoundWeight += node.shardIndices.length;
+					} else if (error instanceof LegallyUnavailableError) {
+						numBlockedWeight += node.shardIndices.length;
+					}
+					throw error;
+				}
+			};
+		});
+
+		const stakingState = await this.stakingState();
+		const numShards = stakingState.n_shards;
+
+		try {
+			const attemptGetMetadata = executors.shift()!;
+			return await attemptGetMetadata();
+		} catch (error) {
+			const queue = new PromiseQueue<BlobMetadataWithId>({
+				maxConcurrency: this.#blobMetadataMaxConcurrencyLimit,
+			});
+
+			return new Promise<BlobMetadataWithId>((resolve, reject) => {
+				let settledCount = 0;
+
+				executors.forEach((executor) => {
+					queue
+						.add(executor)
+						.then((result) => {
+							controller.abort('Blob metadata successfully retrieved.');
+							resolve(result);
+						})
+						.catch((error) => {
+							if (error instanceof UserAbortError) {
+								reject(error);
+							} else if (isQuorum(numBlockedWeight + numNotFoundWeight, numShards)) {
+								if (numNotFoundWeight > numBlockedWeight) {
+									const abortError = new BlobNotCertifiedError(
+										`The specified blob ${blobId} is not certified.`,
+									);
+									controller.abort(abortError);
+									reject(abortError);
+								} else {
+									const abortError = new BlobBlockedError(
+										`The specified blob ${blobId} is blocked.`,
+									);
+									controller.abort(abortError);
+									reject(abortError);
+								}
+							}
+						})
+						.finally(() => {
+							settledCount += 1;
+							if (settledCount === executors.length) {
```

Comment:
You can avoid this logic by making the `forEach` a `map` and then awaiting a `Promise.all` of it that checks if the promise has been settled or not. Seems easier to reason about to me.